lib/src/import_path_base.dart

```diff
@@ -0,0 +1,276 @@
+  List<String> commonRootDirectories = ['web', 'bin', 'src', 'test', 'example'];
+
+  /// Resolves the [searchRoot] using [from] parent.
+  /// See [commonRootDirectories].
+  String resolveSearchRoot() {
+    var rootPath = p.dirname(from.path);
+    var rootDirName = p.split(rootPath).last;
+
+    if (commonRootDirectories.contains(rootDirName)) {
+      var rootPath2 = p.dirname(rootPath);
+      if (rootPath2.isNotEmpty) {
+        rootPath = rootPath2;
+      }
+    }
+
+    var rootUri = from.replace(path: rootPath).toString();
+    return rootUri.endsWith('/') ? rootUri : '$rootUri/';
+  }
+
+  /// Return the search root to [strip] from the displayed import paths.
+  /// If [strip] is `false` returns `null`.
+  /// See [searchRoot].
+  String? get _stripSearchRoot => strip ? searchRoot : null;
+
+  /// Executes the import search and prints the results.
+  /// - [style] defines the output format. Default: [ImportPathStyle.elegant]
+  /// - See [search], [printMessage] and [ASCIIArtTree].
+  Future<ASCIIArtTree?> execute(
+      {ImportPathStyle style = ImportPathStyle.elegant,
+      Directory? packageDirectory}) async {
+    var tree = await search(
+        style: style.asASCIIArtTreeStyle ?? ASCIIArtTreeStyle.elegant,
+        packageDirectory: packageDirectory);
+
+    if (tree != null) {
+      if (!quiet) {
+        printMessage('');
+      }
+
+      if (style == ImportPathStyle.json) {
+        var j = JsonEncoder.withIndent('  ').convert(tree.toJson());
+        printMessage(j);
+      } else {
+        var treeText = tree.generate();
+        printMessage(treeText);
+      }
+    }
+
+    if (!quiet) {
+      if (tree == null) {
+        printMessage(
+            '» Unable to find an import path from $from to $importToFind');
+      } else {
+        var totalFoundPaths = tree.totalLeaves;
+        if (totalFoundPaths > 1) {
+          printMessage(
+              '» Found $totalFoundPaths import paths from $from to $importToFind\n');
+        }
+      }
+    }
+
+    return tree;
+  }
+
+  /// Performs the imports search and returns the tree.
+  /// - [style] defines [ASCIIArtTree] style. Default: [ASCIIArtTreeStyle.elegant]
+  /// - See [searchPaths] and [ASCIIArtTree].
+  Future<ASCIIArtTree?> search(
+      {ASCIIArtTreeStyle style = ASCIIArtTreeStyle.elegant,
+      Directory? packageDirectory}) async {
+    var foundPaths = await searchPaths(packageDirectory: packageDirectory);
+    if (foundPaths.isEmpty) return null;
+
+    var asciiArtTree = ASCIIArtTree.fromPaths(
+      foundPaths,
+      stripPrefix: _stripSearchRoot,
+      style: style,
+    );
+
+    return asciiArtTree;
+  }
+
+  /// Performs the imports search and returns the found import paths.
+  /// - [packageDirectory] is used to resolve . Default: [Directory.current].
+  /// - Uses [ImportPathScanner].
+  Future<List<List<String>>> searchPaths({Directory? packageDirectory}) async {
```

Comment:
Mentioned above but I do really think that List<List<String>> is the wrong type here. Given this is a public API I am a bit hesitant to expose that.

You could type it as `Future<Iterable<Iterable<String>>>` which would give more flexibility to change it to lazy iterables later on?

Reply:
I vote for `Future<Iterable<List<String>>>`, since `List<String>` is a path, won't be possible to be lazy.

Reply:
The path can be lazy though - that is what I am trying to say :). But it cannot be if it is represented by a List.

It would certainly take some work to get there, but fundamentally each path can be created just by doing a traversal of nodes. I am not sure I can fully explain what I have in mind over github comments though lol 🤣 .

We could also just consider actually exposing the graph representation more directly? If you were given the root of the tree, with only the edges that actually lead to the URI that you are searching for present, that would be a compact representation that you could pretty easily convert into anything else.